gcloud/logging/handlers/transports/background_thread.py

```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Transport for Python logging handler that uses a background worker to log
+to Stackdriver Logging asynchronously."""
+
+import atexit
+import copy
+import threading
+
+from gcloud.logging import Client
+from gcloud.logging.handlers.transports.base import Transport
+
+
+class _Worker(object):
+    """ A threaded worker that writes batches of log entires to the logger
+    API.
+
+    This class reuses a single :class:`Batch` method to write successive
+    entries.
+
+    Currently, the only public methods are constructing it (which also starts
+    it) and enqueuing Logger (record, message) pairs.
+    """
+
+    def __init__(self, logger):
+        self.started = False
+        self.stopping = False
+        self.stopped = False
+
+        # _entries_condition is used to signal from the main thread whether
+        # there are any waiting queued logger entries to be written
+        self._entries_condition = threading.Condition()
+
+        # _stop_condition is used to signal from the worker thread to the
+        # main thread that it's finished its last entries
+        self._stop_condition = threading.Condition()
+
+        # This object continually reuses the same :class:`Batch` object to
+        # write multiple entries at the same time.
+        self.logger = logger
+        self.batch = self.logger.batch()
+
+        self._thread = None
+
+        # Number in seconds of  how long to wait for worker to send remaining
+        self._stop_timeout = 5
+
+        self._start()
+
+    def _run(self):
+        """_run is the entry point for the worker thread. It loops
+        until self.stopping is set to true, and commits batch entries
+        written during :meth:`enqueue`"""
+        try:
+            self._entries_condition.acquire()
+            self.started = True
+            while not self.stopping:
+                if len(self.batch.entries) == 0:
+                    # branch coverage of this code extremely flaky
+                    self._entries_condition.wait()  # pragma: NO COVER
```

Comment:
Since this is in a while loop, I'm trying to think of a scenario where this could continue for an indefinite amount of time.  Would this use a ton of CPU if there were a network outage before the logs could be sent?